Fix crash when reaching ScreenSelectMusic with a USB containing songs

CrashCringle12 opened the pull request:
USB song loading logic is handled separately and didn't have the Group/default pack.ini setup applied to it. This PR adds that.

CrashCringle12 committed:
Add Group() for USB songs

teejusb commented on src/Group.cpp:
```diff
@@ -41,11 +41,19 @@ Group::~Group()
+    if (bFromProfile) {
+        sPackIniPath = sDir + "/" + INI_FILE;
+        m_sPath = sDir;
+        m_sGroupName = CUSTOM_SONG_PATH;
```
Oh god don't reference a program level global. Introduce a function in RageUtil to fetch it. Eventually when we clean up the code this will need to change. Same below.

CrashCringle12 replied:
oop updated

CrashCringle12 commented:
Checking something else i just thought of quickly

sukibaby commented on src/RageUtil.cpp:
```diff
@@ -899,6 +899,11 @@ RString custom_songify_path(RString const& path)
 	return CUSTOM_SONG_PATH + parts[parts.size()-2] + "/" + parts[parts.size()-1];
 }
 
+const RString& RageUtil::GetCustomSongPath()
```
the function header is marked `static` in RageUtil.h but not here

teejusb replied:
You don't mark static member functions as static in the implementation file. This is correct

sukibaby commented on src/SongManager.cpp:
```diff
@@ -356,6 +356,32 @@ void SongManager::AddGroup( RString sDir, RString sGroupDirName, Group* group )
+	}
+
+	// Remove the group from its series if the group has one and if the series exists
+	if( group->GetSeries() != "" )
```
```suggestion
	if (!group->GetSeries().empty())

```

teejusb commented:
> Checking something else i just thought of quickly

Okie lmk if things are good before I merge

teejusb commented on src/Profile.cpp:
```diff
@@ -1198,6 +1204,11 @@ void Profile::LoadSongsFromDir(RString const& dir, ProfileSlot prof_slot)
 
 		StripCvsAndSvn(song_folders);
 		StripMacResourceForks(song_folders);
+		RString groupName = RageUtil::GetCustomSongPath();
```
Seems like this groupName isn't used until later. Let's move it closer to it's actual usage below.

teejusb commented on src/SongManager.cpp:
```diff
@@ -356,6 +356,32 @@ void SongManager::AddGroup( RString sDir, RString sGroupDirName, Group* group )
+void SongManager::RemoveGroup( RString sDir, RString sGroupDirName, Group* group )
+{
+	unsigned j;
+	for(j = 0; j < m_sSongGroupNames.size(); ++j)
```
Use braces

teejusb commented on src/SongManager.cpp:
```diff
@@ -356,6 +356,32 @@ void SongManager::AddGroup( RString sDir, RString sGroupDirName, Group* group )
+		series.erase(group);
+	}
+
+	m_sSongGroupBannerPaths.erase(m_sSongGroupBannerPaths.begin() + j);
```
What if the sGroupDirName == m_sSongGroupNames[j] check was never met. Then j would point outside of the size. Add validation checks.

sukibaby replied:
do we need to remove from m_mapNameToGroup as well?

teejusb commented on src/SongManager.cpp:
```diff
@@ -356,6 +356,32 @@ void SongManager::AddGroup( RString sDir, RString sGroupDirName, Group* group )
+			break; 
+		}
+
+	if ( group == nullptr ) {
```
This condition and the next one should be the first two in this function.

sukibaby commented on src/SongManager.cpp:
```diff
@@ -356,6 +356,32 @@ void SongManager::AddGroup( RString sDir, RString sGroupDirName, Group* group )
+		if( sGroupDirName == m_sSongGroupNames[j] ) {
+			break; 
+		}
+
```
you or teejusb may disagree, but i feel it may be safer to find the index of the group we need to remove by using `std::find` so we can know the ensure the group exists and its location is known before proceeding. 

for example
```
index =  std::find(m_sSongGroupNames.begin(), m_sSongGroupNames.end(), sGroupDirName)
```

then you could log a message if the group isn't found,  and use `std::distance` between `m_sSongGroupNames` beginning and store the result in `j`.

doing this would require restructuring the `group->GetSeries()` if statement, but it wouldn't hurt to use `m_mapSeriesToGroups.find(group->GetSeries())`

teejusb replied:
I do prefer std::algorithm functions as long as it doesn't hurt readability. Here while find/distance would totally work, current implementation is also very readable so I dont have a strong preference.

CrashCringle12 commented:
tbh @sukibaby @teejusb Yoinking a lot of that extra "AddGroup()" logic from USB songs since that wasn't something that happened for USB songs previously (I think cause a lot of it is for banner stuff + functions that assume actual groups).  USB songs were pretty isolated and this would technically be changing that somewhat by adding them to mappings they were not previously a part of. Which may or may not be desired later on but trying to keep the PR to just giving usb songs a group and not change behavior of usbs at all.

The only reason i added it to begin with was cause of another bug https://github.com/Simply-Love/Simply-Love-SM5/pull/638 we were looking at earlier that i thought was cause of group stuff but was ultimately discovered to be unrelated and preexisting to this version. Since the bug was ultimately unrelated none of that is really necessary given how USB songs are used and loaded.

Does it make sense for a USB to be part of series? I feel like that doesn't really make sense and is kinda unneeded operations to handle during gameplay. 

With all that being said I removed a lot of this extra logic since it wasn't there before and I kinda don't think it was needed atm.

CrashCringle12 commented:
Other things of note, I removed the addition of GetCustomSongPath() since with 2 USBs it technically wasn't unique. I instead use the ``@mc1`` or ``@mc2`` you get from the dir when USBs are in.

CrashCringle12 commented:
> > Checking something else i just thought of quickly
> 
> Okie lmk if things are good before I merge

lol I got stuck on another unrelated/preexisting bug that i'll put up in a separate issue for later on but its def out of scope for here and not as pressing.

CrashCringle12 committed:
Only check for songs if we're using a memory card

CrashCringle12 committed:
Add additional NULL checks just in case

sukibaby commented:
> tbh @sukibaby @teejusb Yoinking a lot of that extra "AddGroup()" logic from USB songs since that wasn't something that happened for USB songs previously (I think cause a lot of it is for banner stuff + functions that assume actual groups).  USB songs were pretty isolated and this would technically be changing that somewhat by adding them to mappings they were not previously a part of. Which may or may not be desired later on but trying to keep the PR to just giving usb songs a group and not change behavior of usbs at all.
> 
> The only reason i added it to begin with was cause of another bug https://github.com/Simply-Love/Simply-Love-SM5/pull/638 we were looking at earlier that i thought was cause of group stuff but was ultimately discovered to be unrelated and preexisting to this version. Since the bug was ultimately unrelated none of that is really necessary given how USB songs are used and loaded.
> 
> Does it make sense for a USB to be part of series? I feel like that doesn't really make sense and is kinda unneeded operations to handle during gameplay. 
> 
> With all that being said I removed a lot of this extra logic since it wasn't there before and I kinda don't think it was needed atm. 
> 



If it makes everything "just work", or maintains expected behavior as it would be in itgm 0.9 and older, then making USB songs it's own group or series is fine with me. 

fixing pre existing bugs that crop up due to new features is somewhat expectable, in that case it's most important to me to maintain expected behavior, so the change is "transparent" to players. 

Teejusb of course has the final say here but that's my 2 cents on the matter.

teejusb commented on src/Profile.cpp:
```diff
@@ -96,6 +97,12 @@ void Profile::ClearSongs()
 		delete curr_song;
 	}
 	m_songs.clear();
+	LOG->Trace("Profile::ClearSongs()");
```
extra Log->Trace

teejusb committed:
Remove extra Log
